components/board/interrupt_stream.go

```diff
@@ -0,0 +1,111 @@
+package board
+
+import (
+	"context"
+	"sync"
+
+	pb "go.viam.com/api/component/board/v1"
+	"go.viam.com/utils"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+type interruptStream struct {
+	*client
+	streamCancel  context.CancelFunc
+	streamRunning bool
+	streamReady   chan bool
+	streamMu      sync.Mutex
+
+	activeBackgroundWorkers sync.WaitGroup
+	extra                   *structpb.Struct
+}
+
+func (s *interruptStream) startStream(ctx context.Context, interrupts []string, ch chan Tick) error {
+	s.streamMu.Lock()
+	defer s.streamMu.Unlock()
+
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+
+	s.streamRunning = true
+	s.streamReady = make(chan bool)
+	s.activeBackgroundWorkers.Add(1)
+	ctx, cancel := context.WithCancel(ctx)
+	s.streamCancel = cancel
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
+	req := &pb.StreamTicksRequest{
+		Name:     s.client.info.name,
+		PinNames: interrupts,
+		Extra:    s.extra,
+	}
+
+	// This call won't return any errors it had until the client tries to receive.
+	//nolint:errcheck
+	stream, _ := s.client.client.StreamTicks(ctx, req)
+	_, err := stream.Recv()
+	if err != nil {
+		s.client.logger.CError(ctx, err)
+		return err
+	}
+
+	// Create a background go routine to receive from the server stream.
+	utils.ManagedGo(func() {
+		s.recieveFromStream(ctx, stream, ch)
+	},
+		s.activeBackgroundWorkers.Done)
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-s.streamReady:
+		return nil
+	}
+}
+
+func (s *interruptStream) recieveFromStream(ctx context.Context, stream pb.BoardService_StreamTicksClient, ch chan Tick) {
+	defer func() {
+		s.streamMu.Lock()
+		defer s.streamMu.Unlock()
+		s.streamRunning = false
+	}()
+	if s.streamReady != nil {
```

Comment:
comment above this line, "if we're trying to receive already, close the stream then start the for loop again" or something similar.